Filter NULL arguments out when calling non-strict functions/operators

msullivan opened the pull request:
We have two related problems with a related solution:
 - Our implementation of the AND operator, which carefully
   ensures that we return NULL if either input is NULL,
   is a huge performance problem that prevents the use of indexes
 - User defined functions with implementations that aren't strict in
   their arguments can end up producing values when they shouldn't.

To fix this, we add NOT NULL filters when calling functions/operators
with potentially NULL values. To avoid doing this *too* much (and to
make it work more often in singleton mode), we add a notion of
function/operators having a strict implementation, which allows
the check to be skipped. Currently all built-in functions and operators
except for AND and OR are marked as being strict.

Fixes some #3185 related issues.
Fixes #3195.

elprans commented on edb/pgsql/compiler/expr.py:
```diff
@@ -315,6 +315,24 @@ def compile_SliceIndirection(
+            and ref.nullable
+            and typemod == ql_ft.TypeModifier.SingletonType
+        ):
+            raise RuntimeError(
```
Is this a temporary assertion?  If not, then this should be `errors.UnsupportedFeatureError`

msullivan replied:
I made it a RuntimeError because that's what all the similar errors are, but they should probably all be UnsupportedFeatureError

1st1 replied:
Can this surface in a user query? If so, a RuntimeErorr would become an ISE. Maybe we cleanup non-ISE RuntimeError in this file in a separate commit as part of this PR?

elprans commented on edb/schema/functions.py:
```diff
@@ -740,6 +740,9 @@ class CallableObject(
         bool, default=False, inheritable=False, compcoef=0.909)
 
+    impl_is_strict = so.SchemaField(
+        bool, default=True, compcoef=0.4)
```
We have to enforce this by 1) Creating functions with `STRICT` if this is `True`; or 2) For `FROM SQL FUNCTION` functions, validate that the underlying function is `STRICT`.

msullivan replied:
1. Currently we just ignore this on user created functions, though I'll make it go and error instead if it is explicitly specified.

2. Do you know of a way to automatically do that validation?

elprans replied:
See how we use `sql_rval_consistency_check`.  This check would be a lot simpler, you'd only need to query `pg_proc.proisstrict` and `edgedb.raise` if it doesn not match `impl_is_strict`.

elprans replied:
> if it is explicitly specified

It should be impossible to specify in user DDL unless the Field is declared with `allow_ddl_set`, so we're good there.

1st1 commented on edb/edgeql/compiler/func.py:
```diff
@@ -245,6 +245,11 @@ def compile_FunctionCall(
 
+    impl_is_strict = (
+        func.get_impl_is_strict(env.schema)
+        and func.get_language(env.schema) != qlast.Language.EdgeQL
```
Can you add a comment explaining the `and != edgeql` part?

msullivan replied:
Reworked this bit away

1st1 commented on tests/test_edgeql_ddl.py:
```diff
@@ -8898,7 +8896,10 @@ async def test_edgeql_ddl_enum_05(self):
+                 # FIXME: An index on asdf(.color) isn't supported currently
+                 # because passing nullable arguments requires filtering.
+                 # CREATE INDEX ON (asdf(.color));
+                 CREATE INDEX ON (.color);
```
Unless you want to fix this as part of this PR, I'd create an x-fail test so that we don't forget about this.

msullivan replied:
I just fixed it instead

msullivan commented:
I implemented setting strict on functions we define via edgeql or sql, if all the arguments are required singleton arguments. I also added a verification check for all SQL functions used to back functions or operators; this actually found a bug, with `++` on arrays.

I didn't add a verification check for "using SQL operator", because the operators are pretty heavily overloaded and so the check would have needed to check against the argument types also to be useful. This is certainly doable, but I checked the tables manually and the only operator backed by a non-strict function is `||` for arrays, but we actually call `array_cat` on arrays directly anyways, so it didn't seem like it was worth the hassle.

1st1 reviewed (commented):
lgtm

elprans reviewed (approved):
Looks good!
